# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
#выведем общую информацию о таблицах
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 

    display(request.head())
    print(request.info())
    print(f'Таблица "{table}"')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица "books"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица "authors"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица "publishers"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица "ratings"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица "reviews"


In [3]:
#просмотрим данные
books = '''SELECT *
               FROM books
               LIMIT 5'''
pd.io.sql.read_sql(books, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### 1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [4]:
query_1 = '''SELECT COUNT (book_id)
               FROM books
               WHERE publication_date > '2000-01-01' '''
pd.io.sql.read_sql(query_1, con = engine)

count
0    819

Всего после 1 января 2000 года вышло 819 книг.

### 2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [5]:
query_2 = '''
SELECT b.book_id,
       b.title,
       COUNT (DISTINCT re.review_id) AS reviews,
       AVG (r.rating) AS rating_avg
FROM books AS b
LEFT JOIN reviews AS re ON b.book_id=re.book_id
LEFT JOIN ratings AS r ON b.book_id=r.book_id
GROUP BY  b.book_id
ORDER BY reviews DESC
'''
pd.io.sql.read_sql(query_2, con = engine)

book_id                                              title  reviews  \
0        948                            Twilight (Twilight  #1)        7   
1        963                                Water for Elephants        6   
2        734                                   The Glass Castle        6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...        6   
4        695  The Curious Incident of the Dog in the Night-Time        6   
..       ...                                                ...      ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel        0   
996      808                            The Natural Way to Draw        0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites        0   
998      221                          Essential Tales and Poems        0   
999      191  Disney's Beauty and the Beast (A Little Golden...        0   

     rating_avg  
0      3.662500  
1      3.977273  
2      4.206897  
3      4.414634  
4      4.081081  
..          ...  
995    3.666667  
996    3.000000  
997    5.000000  
998    4.000000  
999    4.000000  

[1000 rows x 4 columns]

Количество обзоров и средняя оценка представлены в столбцах reviews и rating_avg соответственно.

### 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [6]:
query_3 = '''
WITH a AS(
SELECT book_id,
       title,
       publisher_id
FROM books
WHERE num_pages > 50)

SELECT p.publisher,
       COUNT (b.book_id) AS total
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
GROUP BY p.publisher
ORDER BY total DESC
LIMIT 1
'''

pd.io.sql.read_sql(query_3, con = engine)

publisher  total
0  Penguin Books     42

Наибольшее количество книг с кол-вом страниц более 50 выпустило издательство Penguin Books.

### 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [7]:
query_4 = '''
WITH a AS(
SELECT b.book_id,
       b.title,
       b.author_id,
       COUNT (r.rating_id) AS total_ratings
FROM books AS b
JOIN ratings AS r ON b.book_id=r.book_id
GROUP BY b.book_id)

SELECT au.author_id,
       au.author,
       AVG (r.rating) AS avg_rating
FROM a
JOIN authors AS au ON au.author_id = a.author_id
JOIN ratings AS r ON a.book_id = r.book_id
WHERE a.total_ratings >= 50
GROUP BY 1,2
ORDER BY avg_rating DESC
LIMIT 1
'''

pd.io.sql.read_sql(query_4, con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

Автором с самой высокой средней оценкой книг, у которых более 50 оценок, является J.K. Rowling/Mary GrandPré	

### 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [8]:
query_5 = '''
SELECT ROUND(AVG(a.total_reviews)) AS avg_reviews
FROM (
SELECT username,
       COUNT(review_id) AS total_reviews
FROM reviews
WHERE username IN (
SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(rating_id) > 50)
GROUP BY username
) AS a
'''

pd.io.sql.read_sql(query_5, con = engine)

avg_reviews
0         24.0

Среднее количество обзоров от пользователей, которые поставили более 50 оценок, составляет 24.

## Вывод
По результатам проекта была проанализирована база данных с информацией о книгах, издательствах, авторах, а также пользовательских обзоров книг. Выявлены следующие факторы:
- Всего после 1 января 2000 года вышло 819 книг, что составляет существенную долю от всех книг;
- Найдены такие величины, как количество обзоров и средняя оценка по кажлой книге, для удобства дальнейшего их анализа;
- Наибольшее количество книг с кол-вом страниц более 50 выпустило издательство Penguin Books;
- Автором с самой высокой средней оценкой книг, у которых более 50 оценок, является J.K. Rowling/Mary GrandPré;
- Среднее количество обзоров от пользователей, которые поставили более 50 оценок, составляет 24.